src/tool/cppwinrt/test/await_adapter.cpp

```diff
@@ -0,0 +1,94 @@
+#include "pch.h"
+#include "winrt/Windows.System.h"
+
+using namespace std::literals;
+using namespace winrt;
+using namespace Windows::Foundation;
+using namespace Windows::System;
+
+namespace
+{
+    static handle signal{ CreateEventW(nullptr, false, false, nullptr) };
+
+    IAsyncAction OtherForegroundAsync()
+    {
+        // Simple coroutine that completes on a unique STA thread.
+
+        auto controller = DispatcherQueueController::CreateOnDedicatedThread();
+        auto dispatcher = controller.DispatcherQueue();
+
+        co_await resume_foreground(dispatcher);
+    }
+
+    IAsyncAction OtherBackgroundAsync()
+    {
+        // Simple coroutine that completes on some MTA thread.
+
+        co_await resume_background();
+    }
+
+    IAsyncAction ForegroundAsync(DispatcherQueue dispatcher)
+    {
+        REQUIRE(!impl::is_sta());
+        co_await resume_foreground(dispatcher);
+        REQUIRE(impl::is_sta());
+
+        // This excercises one STA thread waiting on another thus one context callback
+        // completing on another.
+        uint32_t id = GetCurrentThreadId();
+        co_await OtherForegroundAsync();
+        REQUIRE(id == GetCurrentThreadId());
+
+        // This just avoids the ForegroundAsync coroutine completing before
+        // BackgroundAsync waits on the result, forcing the Completed handler
+        // to be called on the foreground thread. This just makes the test
+        // success/failure more predictable.
+        Sleep(100);
+    }
+
+    fire_and_forget SignalFromForeground(DispatcherQueue dispatcher)
+    {
+        REQUIRE(!impl::is_sta());
+        co_await resume_foreground(dispatcher);
+        REQUIRE(impl::is_sta());
+
+        // Previously, this signal was never raised because the foreground thread
+        // was always blocked waiting for ContextCallback to return.
+        REQUIRE(SetEvent(signal.get()));
+    }
+
+    IAsyncAction BackgroundAsync(DispatcherQueue dispatcher)
+    {
+        // Switch to a background (MTA) thread.
+        co_await resume_background();
+        REQUIRE(!impl::is_sta());
+
+        // This excercises one MTA thread waiting on another and just completing
```

Comment:
This exercises but does not validate. How can we validate that there was no context switch? Should we match up thread IDs?

Reply:
Yes, for the case where we're not queuing we could check the thread Id but that in itself doesn't guarantee a context switch didn't occur since the thread pool may and typically does reuse threads for short-lived work items. I'm not too concerned. The tests are primarily to ensure there's no deadlock and provide a way to manually verify the behavior when desired.

Reply:
Yeah, matching up thread IDs is not sufficient to prove that no context switch occurred. Was not aware what the "standard of testing" is.